src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/IntelliJActivityFactory.java

```diff
@@ -21,7 +21,15 @@ public IntelliJActivityFactory(@NotNull Project project) {
                                                     @NotNull ListenerCallback callback) {
     switch (action) {
       case "openEditor":
-        return new OpenFileListener(callback, arguments.getOrThrow("filepath"), project);
+        return OpenFileListener.create(callback, project, arguments);
+      case "build":
+        return BuildActionListener.create(callback, project,arguments);
+      case "run":
+        return RunActionListener.create(callback, project, arguments);
+      case "classDeclScala":
+        return ClassDeclarationListener.create(callback, project, arguments);
+      case "functionDefinition":
+        return FunctionDefinitionListener.create(callback, project, arguments);
       default:
```

Comment:
Remark: I like this a lot now. We could even allow listeners to be added dynamically (by replacing switch-case with a search in Map<String, ListenerFactory> or such).